src/bgg/circuit/mod.rs

```diff
@@ -192,60 +196,131 @@ impl PolyCircuit {
-            let result = match &gate.gate_type {
-                PolyGateType::Input => {
-                    panic!("Input gate {:?} should already be preloaded", gate);
-                }
-                PolyGateType::Const { digits } => E::from_digits(params, one, &digits),
-                PolyGateType::Add => {
-                    let left = wires[gate.input_gates[0]].as_ref().expect("wire missing for Add");
-                    let right = wires[gate.input_gates[1]].as_ref().expect("wire missing for Add");
-                    left.clone() + right
-                }
-                PolyGateType::Sub => {
-                    let left = wires[gate.input_gates[0]].as_ref().expect("wire missing for Sub");
-                    let right = wires[gate.input_gates[1]].as_ref().expect("wire missing for Sub");
-                    left.clone() - right
-                }
-                PolyGateType::Mul => {
-                    let left = wires[gate.input_gates[0]].as_ref().expect("wire missing for Mul");
-                    let right = wires[gate.input_gates[1]].as_ref().expect("wire missing for Mul");
-                    left.clone() * right
+        debug_mem("Input wires are set");
+
+        for level in levels.iter() {
+            debug_mem("New level started");
+            // All gates in the same level can be processed in parallel.
+            level.par_iter().for_each(|&gate_id| {
+                debug_mem(format!("Gate id {} started", gate_id));
+                if wires.contains_key(&gate_id) {
+                    debug_mem(format!("Gate id {} already evaluated", gate_id));
+                    return;
                 }
-                PolyGateType::Rotate { shift } => {
-                    let input =
-                        wires[gate.input_gates[0]].as_ref().expect("wire missing for Rotate");
-                    input.rotate(params, *shift)
-                }
-                PolyGateType::Call { .. } => {
-                    panic!("no more call gate type during evaluation");
-                }
-            };
-            wires[gate_id] = Some(result);
+                let gate = self.gates.get(&gate_id).expect("gate not found").clone();
+                debug_mem("Get gate");
+                let result = match &gate.gate_type {
+                    PolyGateType::Input => {
+                        panic!("Input gate {:?} should already be preloaded", gate);
+                    }
+                    PolyGateType::Const { digits } => E::from_digits(params, one, &digits),
+                    PolyGateType::Add => {
+                        debug_mem("Add gate start");
+                        let left =
+                            wires.get(&gate.input_gates[0]).expect("wire missing for Add").clone();
+                        let right =
+                            wires.get(&gate.input_gates[1]).expect("wire missing for Add").clone();
+                        let result = left.clone() + right;
```

Comment:
```suggestion
                        let left =
                            wires.get(&gate.input_gates[0]).expect("wire missing for Add").clone();
                        let right =
                            wires.get(&gate.input_gates[1]).expect("wire missing for Add").clone();
                        let result = left + right;
```

probably need to have lhs clone only, same for all gate types